### ISSUE TITLE
Not showing up on iPhone safari

### ISSUE THREAD

**Issue by 9thwall** [foobar404/Wave.js#17]
Hi, 
first off what an amazing script you got here. Great job on this!

I was able to successfully display the waves in firefox and chrome, desktop and android devices. 
But for some reason the waves do not appear on safari iPhone. 

I'm running an icecast server and displaying the waves via audio tag like so..
```
<audio id="radio" crossorigin="anonymous">
       <source src="https://mystation.com/stream" type="audio/mpeg" />
       <source src="https://mystation.com/stream.ogg" type="audio/ogg" />
</audio>
```

and I am triggering the playing of the audio like this..

```
var player = document.getElementById('radio');
player.load();
player.play();
```

I have a canvas html element on the page like this..
`<canvas id="wavesCanvas" height="250" width="500"></canvas>`

The waves is set like this..
```
var wave = new Wave();
var options = {type:"shockwave", colors: ["#6c2a99", "#e01919", "#b4751d"]};
wave.fromElement("radio","wavesCanvas",options);
```

The audio plays on all devices including safari on iPhone.
The wavesjs shows on all browsers _except_ safari on iPhone. 
I should note that I have not tested on safari on a mac because I do not own a mac at the moment. 

Is this a known issue? 
Is there any special configurations needed for this to work on safari?

**Comment by foobar404** [comment]
Yes, this is a known issue. I have tried to fix this in the past but I dont own a apple product so I cant confirm if my changes work. I believe the issue is with the AudioContext instance, which is used to get the frequency information from the audio. Safari wont let you create it, or start an analysis, unless the user has interacted with the window first.

**Comment by julian-weinert** [comment]
Which wouldn't be a problem, because the user interacts with the window by clicking play.

What APIs are vital for Wave.js?
I'm not able to get the example to work with the current desktop Safari. The APIs should be supported. Do you have a small and verbose test-scenario?

**Comment by foobar404** [comment]
You should be able to run "npm run test" in the root of the project and see if waves will render, this does not have error logging though. The API's used are the Canvas API and the Audio Context API. Safari just does not react the same way chrome or firefox does with these API's. The issue is likely with the Audio Context API. Most likely the Analyzer Node is not able to read the frequency information.

**Comment by 9thwall** [comment]
@foobar404 Thank you for the info. 
I will update if I figure out a work around.

**Comment by fffiloni** [comment]
Not showing up in Safari on smartphones, tablets and desktop. 

This link from apple developer may help you @foobar404 to find out how to make it work on Safari ? 
[Safari HTML5 Audio and Video Guide](https://developer.apple.com/library/archive/documentation/AudioVideo/Conceptual/Using_HTML5_Audio_Video/PlayingandSynthesizingSounds/PlayingandSynthesizingSounds.html)

Scroll down to paragraph called "Listing 3-1  A simple synthesizer and audio spectrum visualizer"

**Comment by korbav** [comment]
@fffiloni Do you see any error logged in the developer console?

**Comment by fffiloni** [comment]
I have no errors. Using an audio stream from icecast broadcast give me zeros in the Uint8Array data.
It seems to work with files tho. 

I was able to understand it while playing with this simple fiddle here : [https://jsfiddle.net/6y0dp9e2/](https://jsfiddle.net/6y0dp9e2/) . Fiddle's author wrote that in safari we need to declare audio context after user interaction with audio. 

Using a **file** works fine in the Uint8Array for Safari, but do not with stream or element listening to icecast adress.
——

I found someone who figured out how to make it work using rtc stream, you can read his publication about it [HERE](https://dwayne.xyz/post/audio-visualizations-web-audio-api)

**Comment by korbav** [comment]
> 
> 
> I have no errors. Using an **audio stream from icecast broadcast** give me zeros in the Uint8Array data.
> It seems to work with files tho.
> 
> I was able to understand it while playing with this simple fiddle here : https://jsfiddle.net/6y0dp9e2/ . Fiddle's author wrote that in safari we need to **declare audio analyzer after** user interaction with audio.
> 
> Using a **file** works fine in the Uint8Array for Safari, but do not with stream or element listening to icecast address.
> ——
> 
> I found someone who **figured out how to make it work using rtc stream**, you can read his publication about it [HERE](https://dwayne.xyz/post/audio-visualizations-web-audio-api)
> 
> Hope it helps

Could you check in my repo if the change is fixing your case? [branch stream-problem-safari](https://github.com/korbav/wave.js-typescript/tree/stream-problem-safari)

You just have to clone this branch, install the dependencies and to run `npm run example-from-stream`

**Comment by jailbrokebloke** [comment]
@korbav -- thanks for all your development, bud! -- is there a way to apply your safari fixes to the original wave.js?

**Comment by jailbrokebloke** [comment]
@korbav -- thanks for all your development! 

i tried your rewrite, and unfortunately i still can't get waveforms to work in iOS or desktop Safari :/

**Comment by korbav** [comment]
Hi @jailbrokebloke , thanks for your feedback, then it means that the problem was not related (or at least not only) to the user interaction issue, because this attempt explicitly waits for user interaction to create the Audio Context.
Could you please elaborate a bit on what happens exactly? is there any error/warning, does the browser complain?
About the rewrite, we need to discuss this with @foobar404, I personally think it would be an interesting step for Wave.js to get into it since it brings modern features and probably constitutes a cleaner usage.

**Comment by fffiloni** [comment]
Did you take a look to the publication about rtc sream on Dwayne’s blog ? He gives a reference linked here : https://mdn.github.io/voice-change-o-matic/

This example works on safari, maybe useful to explore the way they built it

**Comment by korbav** [comment]
> 
> 
> Did you take a look to the publication about rtc sream on [Dwayne’s blog](https://dwayne.xyz/post/audio-visualizations-web-audio-api) ? He gives a reference linked here : https://mdn.github.io/voice-change-o-matic/
> 
> This example works on safari, maybe useful to explore the way they built it

I'm gonna give a try to it. Before all, can you confirm that the problem is not related to the access to the microphone? 
Which means you rightly get a prompt asking you if you allow your microphone to be accessed?

**Comment by fffiloni** [comment]
I don't think it's related to the prompt, because using a broadcast link from a webradio for example using the "fromElement" method in wave.js give us no change in the Uint8Array for Safari, but works with a ismple mp3 file. 

I think it can't analyze audio coming from stream for the moment.

**Comment by korbav** [comment]
> 
> 
> I don't think it's related to the prompt, because using a broadcast link from a webradio for example using the "fromElement" method in wave.js give us no change in the Uint8Array for Safari, but works with a simple mp3 file.
> 
> I think it can't analyze audio coming from stream for the moment.

You should create a codesandbox.io demo with your Safari-non-working code. I can't see more without it.

**Comment by olliekav** [comment]
Just adding this for anyone stumbling across this and having issues with Icecast/HLS streams and `createMediaElementSource `. 

This has been an outstanding bug on webkit for ages https://bugs.webkit.org/show_bug.cgi?id=180696.

**Comment by rogeriochaves** [comment]
Hey there 👋
I opened a pull request with a fix for this issue at #84

**Comment by erikmillergalow** [comment]
@rogeriochaves You were able to get data out of the analyser with that code? I've got a page where I wait to create and connect until the user presses a button, the audio plays in all browsers but getByteFrequencyData returns all zeroes in Safari. I'll take a closer look at where our scripts differ
